### PR TITLE
[PTX-20557] Added test scenarios for testing sources option in ha-update

### PR DESCRIPTION
**Pull request by sayalasomayajula-px**
**What this PR does / why we need it**:
This PR tests different scenarios for sources options which will be provided for ha-update for a volume 
We have created two things in this PR

1. A Testcase which will create a base volume and test all scenarios and delete the volume
2. A function called volumeprecheckForHA  which can be used for any volume directly so that it will test all the scenarios for provided volume

These are the scenarios we tested in this PR
Pass ip for id for aggr volumes

[root@ip-10-13-191-13 ~]# /opt/pwx/bin/pxctl v ha-update -r 2 --sources xxxxxxx
Update Volume Replication: Failed to update volume: could not find any pool with id xxxxxxx

pass invalid id:
[root@ip-10-13-191-13 ~]# /opt/pwx/bin/pxctl v ha-update -r 2 --sources 949870d1-e5a5-4f29-9ce4-d84fc63cd6d t1
Update Volume Replication: Failed to update volume: could not find any pool with id 949870d1-e5a5-4f29-9ce4-d84fc63cd6d

pass valid id only for one repl-set and see it does not crash
[root@ip-10-13-191-13 ~]# /opt/pwx/bin/pxctl v ha-update -r 2 --sources 949870d1-e5a5-4f29-9ce4-d84fc63cd6de t1
Update Volume Replication: Replication update started successfully for volume t1

pass valid id not belonging to the volume

[root@ip-10-13-191-13 ~]# /opt/pwx/bin/pxctl v ha-update -r 2 --sources f98ca169-4ae5-4979-b9d4-db27e247a453 t1
Update Volume Replication: Failed to update volume: node f98ca169-4ae5-4979-b9d4-db27e247a453 does not belong to volume's replication set

**Which issue(s) this PR fixes** (optional)
Closes # PTX-20557

**Special notes for your reviewer**:
PASS LOGS:https://jenkins.pwx.dev.purestorage.com/job/Users/job/krishna/job/tp-pxe-byoc/2493/console 

Latest pass log: https://jenkins.pwx.dev.purestorage.com/job/Users/job/krishna/job/tp-pxe-byoc/2714/console

### CODE REVIEW

File: tests/basic/volume_ops_test.go [hunk 1]
```diff
@@ -3202,6 +3202,84 @@ var _ = Describe("{FioClonedVolumeFaultInjection}", func() {
 		}
 	})
 })
+var _ = Describe("{volumeprecheck}", func() {
```
Comment (kphalgun-px) [inline tests/basic/volume_ops_test.go:3205]: Can we use PascalCase?

File: tests/basic/volume_ops_test.go [hunk 2]
```diff
@@ -3202,6 +3202,84 @@ var _ = Describe("{FioClonedVolumeFaultInjection}", func() {
...
+			index := rand.Intn(len(stNodes))
+			selectedNode := &stNodes[index]
+			for _, stNode := range stNodes {
+				if stNode.Name != selectedNode.Name {
+					nodesuuidWithoutReplica = append(nodesuuidWithoutReplica, stNode.Id)
+					nodesIP = append(nodesIP, stNode.Addresses[0])
+				}
+			}
+			id := uuid.New()
+			VolName := fmt.Sprintf("volume_%s", id.String()[:8])
```
Comment (kphalgun-px) [inline tests/basic/volume_ops_test.go:3238]: How about camelCase `volName`?

File: tests/basic/volume_ops_test.go [hunk 3]
```diff
@@ -3202,6 +3202,84 @@ var _ = Describe("{FioClonedVolumeFaultInjection}", func() {
...
+			log.InfoD("Create a volume with a min size on node [%s]", selectedNode.Name)
+			basicVolumeCreate := fmt.Sprintf("volume create --nodes %s %s", selectedNode.Id, VolName)
+			_, err := runPxctlCommand(basicVolumeCreate, *selectedNode, nil)
+			log.FailOnError(err, "volume creation failed on the cluster with volume name [%s]", VolName)
```
Comment (kphalgun-px) [inline tests/basic/volume_ops_test.go:3242]: How about including node details in the log?

File: tests/basic/volume_ops_test.go (hunk 1 above)
Comment (sayalasomayajula-px) [reply tests/basic/volume_ops_test.go]: changed

File: tests/basic/volume_ops_test.go (hunk 2 above)
Comment (sayalasomayajula-px) [reply tests/basic/volume_ops_test.go]: changed it

File: tests/basic/volume_ops_test.go (hunk 3 above)
Comment (sayalasomayajula-px) [reply tests/basic/volume_ops_test.go]: added node details

File: tests/basic/volume_ops_test.go [hunk 4]
```diff
@@ -3202,6 +3202,85 @@ var _ = Describe("{FioClonedVolumeFaultInjection}", func() {
...
+			id := uuid.New()
+			volName := fmt.Sprintf("volume_%s", id.String()[:8])
+			log.InfoD("Create a volume with a min size on node [%s]", selectedNode.Name)
+			basicVolumeCreate := fmt.Sprintf("volume create --nodes %s %s", selectedNode.Id, volName)
+			_, err := runPxctlCommand(basicVolumeCreate, *selectedNode, nil)
+			log.FailOnError(err, "volume creation failed on the cluster with volume name [%s] on node %s", volName, selectedNode.Name)
+			log.InfoD("Base Volume creation with volume name %s successful", volName)
+
+			log.InfoD("Test the ha-update sources option with a uuid of other nodes on which node is not present")
+			wrongUuidcmd := fmt.Sprintf("v ha-update %s --repl 2 --sources %s", volName, nodesuuidWithoutReplica[rand.Intn(len(nodesuuidWithoutReplica))])
```
Comment (lsrinivas-pure) [inline tests/basic/volume_ops_test.go:3246]: test with aggregated volumes

File: tests/common.go [hunk 5]
```diff
@@ -11569,3 +11569,64 @@ func SplitStorageDriverUpgradeURL(upgradeURL string) (string, string, error) {
 	pxVersion := pathSegments[len(pathSegments)-1]
 	return endpoint.String(), pxVersion, nil
 }
+
+func volumePreCheckForHA(v volume.Volume) error {
```
Comment (lsrinivas-pure) [inline tests/common.go:11573]: where is this function used?

**Review by lsrinivas-pure — CHANGES_REQUESTED** [review]

File: tests/basic/volume_ops_test.go (hunk 4 above)
Comment (sayalasomayajula-px) [reply tests/basic/volume_ops_test.go:3251]: https://jenkins.pwx.dev.purestorage.com/job/Users/job/krishna/job/tp-pxe-byoc/2629/console. -- tested with aggregated volumes , pass log

File: tests/common.go (hunk 5 above)
Comment (sayalasomayajula-px) [reply tests/common.go:11573]: removing it

File: tests/basic/volume_ops_test.go (hunk 4 above)
Comment (lsrinivas-pure) [reply tests/basic/volume_ops_test.go:3251]: parameteize the aggregation and replication base on number of storage nodes and  create volume using replication and aggregation combined

Comment (sayalasomayajula-px) [reply tests/basic/volume_ops_test.go]: added it ,got the pass logs https://jenkins.pwx.dev.purestorage.com/job/Users/job/krishna/job/tp-pxe-byoc/2676/console

File: tests/basic/volume_ops_test.go [hunk 6]
```diff
@@ -3202,6 +3202,108 @@ var _ = Describe("{FioClonedVolumeFaultInjection}", func() {
...
+			for _, stNode := range stNodes {
+				if stNode.Name != selectedNode.Name {
+					nodesuuidWithoutReplica = append(nodesuuidWithoutReplica, stNode.Id)
+					nodesIP = append(nodesIP, stNode.Addresses[0])
+				}
+			}
+			var aggr_level int
+			var repl_level int
+			storageNodes := node.GetStorageNodes()
+			if len(storageNodes) == 4 && len(storageNodes) < 9 {
```
Comment (lsrinivas-pure) [inline tests/basic/volume_ops_test.go:3240]: shouldn't this be >=4 and < 9

File: tests/basic/volume_ops_test.go [hunk 7]
```diff
@@ -3202,6 +3202,108 @@ var _ = Describe("{FioClonedVolumeFaultInjection}", func() {
...
+				log.InfoD("Setting the aggr_level and repl_level to 2 as storage nodes in the cluster are %d", len(storageNodes))
+				aggr_level = 2
+				repl_level = 2
+			} else if len(storageNodes) >= 9 {
+				log.InfoD("Setting the aggr_level and repl_level to 2 as storage nodes in the cluster are %d", len(storageNodes))
+				aggr_level = 3
+				repl_level = 3
+			} else {
+				log.InfoD("Setting the aggr_level and repl_level to 1 as storage nodes in the cluster are %d", len(storageNodes))
+				aggr_level = 1
```
Comment (lsrinivas-pure) [inline tests/basic/volume_ops_test.go:3250]: make aggr_level-2

File: tests/basic/volume_ops_test.go [hunk 8]
```diff
@@ -3202,6 +3202,108 @@ var _ = Describe("{FioClonedVolumeFaultInjection}", func() {
...
+		Step(stepLog, func() {
+			log.InfoD(stepLog)
+			stNodes := node.GetStorageDriverNodes()
+			nodesuuidWithoutReplica := make([]string, 0)
+			nodesIP := make([]string, 0)
+			index := rand.Intn(len(stNodes))
+			selectedNode := &stNodes[index]
+			for _, stNode := range stNodes {
+				if stNode.Name != selectedNode.Name {
+					nodesuuidWithoutReplica = append(nodesuuidWithoutReplica, stNode.Id)
```
Comment (lsrinivas-pure) [inline tests/basic/volume_ops_test.go:3233]: if u have r 2 or r 3 this logic will fail. please have a look

**Review by lsrinivas-pure — CHANGES_REQUESTED** [review]

File: tests/basic/volume_ops_test.go [hunk 9]
```diff
@@ -3202,6 +3202,120 @@ var _ = Describe("{FioClonedVolumeFaultInjection}", func() {
...
+			nodesIP := make([]string, 0)
+
+			index := rand.Intn(len(stNodes))
+			selectedNode := &stNodes[index]
+
+			var aggr_level int
+			var repl_level int
+			storageNodes := node.GetStorageNodes()
+			if len(storageNodes) >= 4 && len(storageNodes) < 9 {
+				log.InfoD("Setting the aggr_level and repl_level to 2 as storage nodes in the cluster are %d", len(storageNodes))
```
Comment (lsrinivas-pure) [inline tests/basic/volume_ops_test.go:3238]: nit: you can print this log at once after setting aggr_level and repl_level values

**Review by lsrinivas-pure — APPROVED** [review]
LGTM

**Pull request merged**